src/main/java/net/neoforged/neoforge/capabilities/CapabilityHooks.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.List;
+import net.minecraft.core.Direction;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.WorldlyContainerHolder;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.animal.horse.AbstractHorse;
+import net.minecraft.world.item.BucketItem;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.ChestBlock;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.neoforge.common.NeoForgeMod;
+import net.neoforged.neoforge.event.TickEvent;
+import net.neoforged.neoforge.event.level.ChunkEvent;
+import net.neoforged.neoforge.fluids.capability.wrappers.FluidBucketWrapper;
+import net.neoforged.neoforge.items.VanillaHopperItemHandler;
+import net.neoforged.neoforge.items.wrapper.CombinedInvWrapper;
+import net.neoforged.neoforge.items.wrapper.EntityArmorInvWrapper;
+import net.neoforged.neoforge.items.wrapper.EntityHandsInvWrapper;
+import net.neoforged.neoforge.items.wrapper.ForwardingItemHandler;
+import net.neoforged.neoforge.items.wrapper.InvWrapper;
+import net.neoforged.neoforge.items.wrapper.PlayerInvWrapper;
+import net.neoforged.neoforge.items.wrapper.ShulkerItemStackInvWrapper;
+import net.neoforged.neoforge.items.wrapper.SidedInvWrapper;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public class CapabilityHooks {
+    private static boolean initialized = false;
+    static boolean initFinished = false;
+
+    public static void init() {
+        if (initialized)
+            throw new IllegalArgumentException("CapabilityHooks.init() called twice");
+        initialized = true;
+
+        var event = new RegisterCapabilitiesEvent();
+        ModLoader.get().postEventWrapContainerInModOrder(event);
+
+        initFinished = true;
+    }
+
+    public static void registerVanillaProviders(RegisterCapabilitiesEvent event) {
+        // Blocks
+        var composterBlock = (WorldlyContainerHolder) Blocks.COMPOSTER;
+        event.registerBlock(Capabilities.ItemHandler.BLOCK, (level, pos, state, blockEntity, side) -> {
+            // Return a wrapper that gets re-evaluated every time it is accessed
+            // Invalidation is taken care of by the patches to ComposterBlock
+            if (side == null)
+                return new ForwardingItemHandler(() -> new InvWrapper(composterBlock.getContainer(state, level, pos)));
+            else
+                return new ForwardingItemHandler(() -> new SidedInvWrapper(composterBlock.getContainer(state, level, pos), side));
```

Comment:
The current solution to this has a few bugs when it comes to `BlockCapabilityCache` and a fundamental flaw in how the `ForwardingItemHandler` functions for the composter:
- Capability is queried
- The supplier passed to the `ForwardingItemHandler` captures the `state`, `level`, and `pos`
- Composter uses the state for determining which container to return

Because the state was captured it never actually is getting updated when the composter's container is being resolved... which means that the `BlockCapabilityCache`'s cached `ForwardingItemHandler` will always use the state the composter had when first queried, and thus be using the wrong container some of the time.

Reply:
I added a blockstate supplier to query the level each time, that should hopefully fix it?

Reply:
Tested and it does seem to fix it. In an ideal world we could somehow cache the state between operations (so that we don't have to requery the world for each `getStackInSlot`, `insertItem` (simulated and execution) and other methods, and could instead somehow know that until an executing insert/extract call within the single tick it is fine to use the same state.. But there isn't a great way to do that without lots of added complexity.

Before I mark this as resolved though, do we potentially want to add to `IBlockCapabilityProvider` some sort of note/gotcha to point out to modders that depending on how they do their handlers bear in mind that it can be cached so extra caution should be taken when it is based on the state and invalidation does not happen on every state change?

Reply:
Would be a good idea to add a disclaimer, but I am not sure how to best phrase it without making the documentation too long. Maybe as an example in the class javadoc of `BlockCapability.java`.

Reply:
Added a `Be careful that if you don't invalidate for every state change, you should not capture the state parameter because the state might change!` disclaimer.